File: Packages/NMP/Sources/NMP/Session.swift
import Foundation
import NMPFFI

/// One validated decoded Nostr public key.
public struct NMPPublicKey: @unchecked Sendable, Equatable {
    let ffi: FfiPublicKey

    public init(bytes: Data) throws {
        ffi = try nmpRethrowing { try FfiPublicKey.fromBytes(bytes: bytes) }
    }

    init(ffi: FfiPublicKey) {
        self.ffi = ffi
    }

    public var bytes: Data { ffi.bytes() }

    public static func == (lhs: Self, rhs: Self) -> Bool {
        lhs.bytes == rhs.bytes
    }
}

/// One validated decoded Nostr private key.
///
/// The Swift wrapper provides no byte accessor and renders only a redacted
/// description. NMP wipes the native buffer it owns when released; it makes
/// no claim about the caller's original `Data` or FFI transport copies.
public final class NMPPrivateKey: @unchecked Sendable, CustomStringConvertible,
    CustomDebugStringConvertible
{
    let ffi: FfiPrivateKey

    public init(bytes: Data) throws {
        ffi = try nmpRethrowing { try FfiPrivateKey.fromBytes(bytes: bytes) }
    }

    private init(ffi: FfiPrivateKey) {
        self.ffi = ffi
    }

    public static func generate() -> NMPPrivateKey {
        NMPPrivateKey(ffi: FfiPrivateKey.generate())
    }

    public var description: String { "NMPPrivateKey(<redacted>)" }
    public var debugDescription: String { description }
}

/// Opaque bytes representing one complete engine session.
///
/// The bytes may be stored and restored as a whole, but their account and
/// signer-provider material is owned and interpreted only by NMP. Treat them
/// as sensitive. String interpolation and debug output are always redacted.
public final class NMPSessionPayload: @unchecked Sendable, CustomStringConvertible,
    CustomDebugStringConvertible
{
    let ffi: FfiSessionPayload

    public init(bytes: Data) {
        ffi = FfiSessionPayload.fromBytes(bytes: bytes)
    }

    init(ffi: FfiSessionPayload) {
        self.ffi = ffi
    }

    /// The opaque transport bytes. Store and restore this value atomically;
    /// do not parse or partially modify it.
    public var bytes: Data { ffi.bytes() }

    public var description: String { "NMPSessionPayload(<redacted>)" }
    public var debugDescription: String { description }
}

/// The persistable signer-provider kind configured for an account.
/// Availability is deliberately not inferred from this value.
public enum NMPSessionProviderKind: Sendable, Equatable {
    case localKey

    init(_ ffi: FfiSessionProviderKind) {
        switch ffi {
        case .localKey: self = .localKey
        }
    }
}

/// The current operational state of one cryptographic capability.
public enum NMPCapabilityAvailability: Sendable, Equatable {
    case unsupported
    case available
    case unavailable(reason: String)

    init(_ ffi: FfiCapabilityAvailability) {
        switch ffi {
        case .unsupported: self = .unsupported
        case .available: self = .available
        case .unavailable(let reason): self = .unavailable(reason: reason)
        }
    }
}

/// One account in the engine's current session.
///
/// Signer-backed and public-key-only accounts use this same handle. Removing
/// the handle removes the whole account, including its provider. Account
/// identity is its public key; re-adding a provider updates that account.
public final class NMPSessionAccount: @unchecked Sendable {
    public let publicKey: NMPPublicKey
    public let providerKind: NMPSessionProviderKind?
    public let signingAvailability: NMPCapabilityAvailability
    let ffi: FfiSessionAccount

    init(ffi: FfiSessionAccount) {
        self.ffi = ffi
        publicKey = NMPPublicKey(ffi: ffi.publicKey())
        providerKind = ffi.provider().map(NMPSessionProviderKind.init)
        signingAvailability = NMPCapabilityAvailability(ffi.signingAvailability())
    }
}

/// The engine-owned view of one complete account session.
///
/// This object does not own another runtime or lifecycle. It projects session
/// mutations through its `NMPEngine` and never retains secret material itself.
public final class NMPSession: @unchecked Sendable {
    private let ffi: NmpEngineProtocol

    init(ffi: NmpEngineProtocol) {
        self.ffi = ffi
    }

    public var accounts: [NMPSessionAccount] {
        get throws {
            try nmpRethrowing {
                try ffi.session().accounts.map(NMPSessionAccount.init)
            }
        }
    }

    public var current: NMPSessionAccount? {
        get throws {
            try nmpRethrowing {
                let snapshot = try ffi.session()
                guard let currentPublicKey = snapshot.currentPublicKey else {
                    return nil
                }
                return snapshot.accounts
                    .first { $0.publicKey().bytes() == currentPublicKey.bytes() }
                    .map(NMPSessionAccount.init)
            }
        }
    }

    /// Add a private-key-backed account. When requested, selection happens in
    /// the same engine transition as addition.
    public func add(
        privateKey: NMPPrivateKey,
        makeCurrent: Bool = false
    ) throws -> NMPSessionAccount {
        try nmpRethrowing {
            NMPSessionAccount(
                ffi: try ffi.addPrivateKeyAccount(
                    privateKey: privateKey.ffi,
                    makeCurrent: makeCurrent
                )
            )
        }
    }

    /// Add an ordinary account without a signer provider. Decode bech32 at
    /// the app's human input boundary before calling this method.
    public func add(
        publicKey: NMPPublicKey,
        makeCurrent: Bool = false
    ) throws -> NMPSessionAccount {
        try nmpRethrowing {
            NMPSessionAccount(
                ffi: try ffi.addPublicKeyAccount(
                    publicKey: publicKey.ffi,
                    makeCurrent: makeCurrent
                )
            )
        }
    }

    public func makeCurrent(_ account: NMPSessionAccount) throws {
        try nmpRethrowing {
            try ffi.makeCurrentAccount(account: account.ffi)
        }
    }

    /// Remove the whole account at this public key. Repeated removal is a no-op.
    @discardableResult
    public func remove(_ account: NMPSessionAccount) throws -> Bool {
        try nmpRethrowing {
            try ffi.removeAccount(account: account.ffi)
        }
    }

    /// Clear accounts, providers, and current selection without clearing
    /// cached events, receipts, or accepted write obligations.
    public func clear() throws {
        try nmpRethrowing { try ffi.clearSession() }
    }

    /// Export the one complete opaque value suitable for atomic app storage.
    public func export() throws -> NMPSessionPayload {
        try nmpRethrowing {
            NMPSessionPayload(ffi: try ffi.exportSession())
        }
    }
}

File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/Session.kt
package com.nmp.sdk

import uniffi.nmp_ffi.FfiCapabilityAvailability
import uniffi.nmp_ffi.FfiPrivateKey
import uniffi.nmp_ffi.FfiPublicKey
import uniffi.nmp_ffi.FfiSessionAccount
import uniffi.nmp_ffi.FfiSessionPayload
import uniffi.nmp_ffi.FfiSessionProviderKind
import uniffi.nmp_ffi.NmpEngine

/**
 * Opaque bytes representing one complete engine session.
 *
 * Store and restore the bytes as a whole. Account and provider restoration
 * material belongs to NMP and must not be parsed, partially edited, logged,
 * or rendered by the app.
 */
class NMPSessionPayload internal constructor(
    internal val ffi: FfiSessionPayload,
) {
    constructor(bytes: ByteArray) : this(FfiSessionPayload.fromBytes(bytes))

    /** A defensive copy of the opaque, sensitive transport bytes. */
    fun bytes(): ByteArray = ffi.bytes()

    override fun toString(): String = "NMPSessionPayload(<redacted>)"
}

/** A decoded Nostr public key. Bech32 exists only at the app's human boundary. */
class NMPPublicKey private constructor(
    internal val ffi: FfiPublicKey,
) {
    constructor(bytes: ByteArray) : this(
        nmpRethrowing { FfiPublicKey.fromBytes(bytes.copyOf()) },
    )

    /** A defensive copy of the exact decoded 32-byte key. */
    val bytes: ByteArray
        get() = ffi.bytes()

    override fun equals(other: Any?): Boolean =
        other is NMPPublicKey && bytes.contentEquals(other.bytes)

    override fun hashCode(): Int = bytes.contentHashCode()

    override fun toString(): String = "NMPPublicKey"

    internal companion object {
        fun from(ffi: FfiPublicKey): NMPPublicKey = NMPPublicKey(ffi)
    }
}

/**
 * A decoded private key handed directly to NMP's provider boundary.
 *
 * It deliberately has no byte accessor and its string representation is
 * redacted. NMP owns the provider after account addition.
 */
class NMPPrivateKey private constructor(
    internal val ffi: FfiPrivateKey,
) {
    constructor(bytes: ByteArray) : this(
        nmpRethrowing { FfiPrivateKey.fromBytes(bytes) },
    )

    override fun toString(): String = "NMPPrivateKey(<redacted>)"

    companion object {
        /** Generate one decoded private key inside NMP's native boundary. */
        fun generate(): NMPPrivateKey = NMPPrivateKey(FfiPrivateKey.generate())
    }
}

/** The configured provider kind, independent of its current availability. */
enum class NMPSessionProviderKind {
    LocalKey,
    ;

    internal companion object {
        fun from(ffi: FfiSessionProviderKind): NMPSessionProviderKind =
            when (ffi) {
                FfiSessionProviderKind.LOCAL_KEY -> LocalKey
            }
    }
}

/** The current operational state of one cryptographic capability. */
sealed interface NMPCapabilityAvailability {
    data object Unsupported : NMPCapabilityAvailability

    data object Available : NMPCapabilityAvailability

    data class Unavailable(val reason: String) : NMPCapabilityAvailability

    companion object {
        internal fun from(ffi: FfiCapabilityAvailability): NMPCapabilityAvailability =
            when (ffi) {
                is FfiCapabilityAvailability.Unsupported -> Unsupported
                is FfiCapabilityAvailability.Available -> Available
                is FfiCapabilityAvailability.Unavailable -> Unavailable(ffi.reason)
            }
    }
}

/**
 * Opaque identity for one exact account in the current engine session.
 *
 * Signer-backed and public-key-only accounts use this same handle. Removing
 * it removes the whole account, including its provider. Repeated removal is a
 * no-op; adding the same public key again refers to that same account identity.
 */
class NMPSessionAccount internal constructor(
    internal val ffi: FfiSessionAccount,
) {
    val publicKey: NMPPublicKey = NMPPublicKey.from(ffi.publicKey())
    val providerKind: NMPSessionProviderKind? =
        ffi.provider()?.let(NMPSessionProviderKind::from)
    val signingAvailability: NMPCapabilityAvailability =
        NMPCapabilityAvailability.from(ffi.signingAvailability())

    override fun toString(): String = "NMPSessionAccount"
}

/**
 * The engine-owned view of one complete account session.
 *
 * This object owns no second runtime. All mutations go through the same
 * engine transition, and no secret or restoration material is retained in
 * this Kotlin wrapper.
 */
class NMPSession internal constructor(
    private val ffi: NmpEngine,
) {
    val accounts: List<NMPSessionAccount>
        get() = nmpRethrowing { ffi.session().accounts.map(::NMPSessionAccount) }

    val current: NMPSessionAccount?
        get() =
            nmpRethrowing {
                val snapshot = ffi.session()
                val currentPublicKey = snapshot.currentPublicKey ?: return@nmpRethrowing null
                snapshot.accounts
                    .firstOrNull {
                        it.publicKey().bytes().contentEquals(currentPublicKey.bytes())
                    }
                    ?.let(::NMPSessionAccount)
            }

    /** Add a private-key-backed account, optionally selecting it atomically. */
    fun add(
        privateKey: NMPPrivateKey,
        makeCurrent: Boolean = false,
    ): NMPSessionAccount =
        nmpRethrowing {
            NMPSessionAccount(ffi.addPrivateKeyAccount(privateKey.ffi, makeCurrent))
        }

    /**
     * Add an ordinary account without a signer provider.
     *
     * Decode bech32 at the app's human input boundary before constructing the
     * decoded [NMPPublicKey] passed here.
     */
    fun add(
        publicKey: NMPPublicKey,
        makeCurrent: Boolean = false,
    ): NMPSessionAccount =
        nmpRethrowing {
            NMPSessionAccount(ffi.addPublicKeyAccount(publicKey.ffi, makeCurrent))
        }

    fun makeCurrent(account: NMPSessionAccount) =
        nmpRethrowing { ffi.makeCurrentAccount(account.ffi) }

    /** Remove this whole account. Repeated removal is a no-op. */
    fun remove(account: NMPSessionAccount): Boolean =
        nmpRethrowing { ffi.removeAccount(account.ffi) }

    /**
     * Clear accounts, providers, and current selection without clearing
     * cached events, receipts, or accepted write obligations.
     */
    fun clear() = nmpRethrowing { ffi.clearSession() }

    /** Export the one complete opaque value for atomic app storage. */
    fun export(): NMPSessionPayload =
        nmpRethrowing { NMPSessionPayload(ffi.exportSession()) }
}
